util/symbolize/dwarf3ranges.cc

```diff
@@ -47,4 +60,199 @@ void AddressRangeList::ReadRangeList(uint64 offset, uint64 base,
   } while (true);
 }
 
+
+void AddressRangeList::ReadDwarfRngListsDirectly(uint64 offset, uint64 base,
+                                            AddressRangeList::RangeList* ranges, uint64 addr_base) {
+  // Read rnglists from .debug_rnglists when the format was DW_FORM_sec_offset.
+  // Since we are calculating the address with DW_FORM_sec_offset, 
+  //  we don't care about (or need) the index to rngdatamap_ here. 
+  //  Therefore, we are getting the first element (rngdatamap_.begin())
+  //  and getting the address_size from that element. 
+  //  (The address size is equal for all elements.)
+  auto i = rngdatamap_.begin();
+  CHECK(i != rngdatamap_.end());
+  ReadDwarfRngLists(base, ranges, buffer_ + offset, addr_base, i->second.address_size);
+}
+
+void AddressRangeList::ReadDwarfRngListwithOffsetArray(uint64 offset, uint64 base,
+                                            AddressRangeList::RangeList* ranges, uint64 addr_base, uint64 range_base_) {
+  auto i = rngdatamap_.find(range_base_);
+  CHECK(i != rngdatamap_.end());
+  RngListsData& rnglistsdata = i->second;                                              
+  ReadDwarfRngLists(base, ranges, buffer_ + rnglistsdata.rnglist_base_ + offset, addr_base, rnglistsdata.address_size);
+}
+
+void AddressRangeList::ReadDwarfRngLists(uint64 base,
+                                         AddressRangeList::RangeList* ranges,
+                                         const char* pos, uint64 addr_base,
+                                         uint8 address_size) {
+
+    CHECK(is_rnglists_section);
+
+    bool read_next_entry = true;
+
+    do {
+        CHECK(pos < buffer_ + buffer_length_);
+        uint8 entry_kind = reader_->ReadOneByte(pos); pos += 1;
+
+        switch (entry_kind) {
+          case DW_RLE_end_of_list: {
+            read_next_entry = false;
+            break;
+          }
+          case DW_RLE_base_addressx: {
+            size_t len = 0;
+            uint64 addr_section_address = reader_->ReadUnsignedLEB128(pos, &len); pos += len;
+            CHECK(addr_buffer_ + addr_base + addr_section_address * reader_->AddressSize()
+                   <= addr_buffer_ + addr_buffer_length_);
+            base = reader_->ReadAddress(addr_buffer_ + addr_base + addr_section_address * reader_->AddressSize());
+            break;
+          }
+          case DW_RLE_startx_endx: {
+            size_t len = 0;
+            // Start
+            uint64 start_index = reader_->ReadUnsignedLEB128(pos, &len); pos += len;
+            CHECK(pos <= (buffer_ + buffer_length_));
```

Comment:
The parenthesis around `buffer_ + buffer_length_` are unnecessary and inconsistent with other places. Please remove them here and elsewhere.